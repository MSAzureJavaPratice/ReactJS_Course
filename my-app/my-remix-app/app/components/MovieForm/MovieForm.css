.movie-form {
    display: flex;
    flex-direction: column;
    gap: 15px;
}

.movie-form label {
    display: flex;
    flex-direction: column;
    font-weight: bold;
}

.movie-form button {
    padding: 10px;
    background-color: red;
    color: white;
    border: none;
    border-radius: 4px;
    cursor: pointer;
}