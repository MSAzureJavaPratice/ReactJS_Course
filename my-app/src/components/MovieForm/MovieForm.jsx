import React, { useEffect, useState } from "react";
import { useForm, Controller } from "react-hook-form";
import "./MovieForm.css";

const availableGenres = [
  "Action",
  "Crime",
  "Drama",
  "Sci-Fi",
  "Thriller",
  "Fantasy",
  "Adventure",
  "Romance",
];

const MovieForm = ({ initialMovieInfo = {}, onSubmit }) => {
  const [movieId, setMovieId] = useState(initialMovieInfo.id || "");

  const {
    register,
    handleSubmit,
    control,
    reset,
    formState: { errors },
  } = useForm({
    defaultValues: {
      title: initialMovieInfo.title || "",
      release_date: initialMovieInfo.release_date || "",
      genres: initialMovieInfo.genres || [],
      runtime: initialMovieInfo.runtime || "",
      overview: initialMovieInfo.overview || "",
      vote_average: initialMovieInfo.vote_average || "",
      poster_path: initialMovieInfo.poster_path || "",
    },
  });

  useEffect(() => {
    // Update form values and movieId if initialMovieInfo changes
    reset({
      title: initialMovieInfo.title || "",
      release_date: initialMovieInfo.release_date || "",
      genres: initialMovieInfo.genres || [],
      runtime: initialMovieInfo.runtime || "",
      overview: initialMovieInfo.overview || "",
      vote_average: initialMovieInfo.vote_average || "",
      poster_path: initialMovieInfo.poster_path || "",
    });
    setMovieId(initialMovieInfo.id || "");
  }, [initialMovieInfo, reset]);

  const submitHandler = (data) => {
    onSubmit({
      ...data,
      id: movieId, // Include the movie ID in the submitted data
      runtime: parseInt(data.runtime, 10),
      vote_average: parseFloat(data.vote_average),
    });
  };

  return (
    <form className="movie-form" onSubmit={handleSubmit(submitHandler)}>
      <div>
        <label>
          Title:
          <input {...register("title", { required: "Title is required" })} />
          {errors.title && <span className="error">{errors.title.message}</span>}
        </label>
      </div>
      <div>
        <label>
          Release Date:
          <input
            type="date"
            {...register("release_date", { required: "Release date is required" })}
          />
          {errors.release_date && (
            <span className="error">{errors.release_date.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Genre:
          <Controller
            control={control}
            name="genres"
            rules={{ required: "At least one genre must be selected" }}
            render={({ field: { value, onChange } }) => (
              <select
                multiple
                value={value || []}
                onChange={(e) => {
                  const selectedOptions = Array.from(e.target.selectedOptions, (option) => option.value);
                  onChange(selectedOptions);
                }}
              >
                {availableGenres.map((genre) => (
                  <option key={genre} value={genre}>
                    {genre}
                  </option>
                ))}
              </select>
            )}
          />
        </label>
        {errors.genres && <span className="error">{errors.genres.message}</span>}
      </div>
      <div>
        <label>
          Runtime:
          <input
            type="number"
            {...register("runtime", {
              required: "Runtime is required",
              min: { value: 1, message: "Runtime must be at least 1 minute" },
            })}
          />
          {errors.runtime && (
            <span className="error">{errors.runtime.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Rating:
          <input
            type="number"
            step="0.1"
            {...register("vote_average", {
              required: "Rating is required",
              min: { value: 0, message: "Rating must be at least 0" },
              max: { value: 10, message: "Rating cannot exceed 10" },
            })}
          />
          {errors.vote_average && (
            <span className="error">{errors.vote_average.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Poster URL:
          <input
            {...register("poster_path", {
              required: "Poster URL is required",
              pattern: {
                value: /^(https?:\/\/.*\.(?:png|jpg|jpeg|gif|svg))$/i,
                message: "Enter a valid image URL",
              },
            })}
          />
          {errors.poster_path && (
            <span className="error">{errors.poster_path.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Overview:
          <textarea
            rows="5"
            {...register("overview", {
              required: "Overview is required",
              minLength: { value: 10, message: "Overview must be at least 10 characters" },
            })}
          />
          {errors.overview && (
            <span className="error">{errors.overview.message}</span>
          )}
        </label>
      </div>
      <button type="submit">Submit</button>
    </form>
  );
};

export default MovieForm;