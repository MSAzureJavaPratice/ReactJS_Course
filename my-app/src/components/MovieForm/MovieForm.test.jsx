import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import MovieForm from "./MovieForm";

describe("MovieForm Component", () => {
  const onSubmit = jest.fn();

  test("renders empty form when no initialMovieInfo is provided", () => {
    render(<MovieForm onSubmit={onSubmit} />);
    expect(screen.getByLabelText("Title:")).toHaveValue("");
    expect(screen.getByLabelText("Release Date:")).toHaveValue(""); // Fixed label text
    expect(screen.getByLabelText("Genre:")).toHaveValue([]); // For multi-select, it should be an empty array
    expect(screen.getByLabelText("Overview:")).toHaveValue(""); // Corrected description to overview
  });

  test("renders form with initial values when initialMovieInfo is provided", () => {
    const initialMovieInfo = {
      title: "Inception",
      release_date: "2010-07-16", // Correct the date format
      genres: ["Sci-Fi"], // Ensure this is an array since it's a multi-select
      overview: "A movie about dreams.",
    };

    render(
      <MovieForm initialMovieInfo={initialMovieInfo} onSubmit={onSubmit} />
    );
    expect(screen.getByLabelText("Title:")).toHaveValue("Inception");
    expect(screen.getByLabelText("Release Date:")).toHaveValue("2010-07-16");
    expect(screen.getByLabelText("Genre:")).toHaveValue(["Sci-Fi"]); // For multi-select, check the set values
    expect(screen.getByLabelText("Overview:")).toHaveValue(
      "A movie about dreams."
    );
  });

  test("calls onSubmit with form data when submitted", () => {
    const initialMovieInfo = {
      title: "Inception",
      release_date: "2010-07-16",
      genres: ["Sci-Fi"],
      overview: "A movie about dreams.",
    };

    render(
      <MovieForm initialMovieInfo={initialMovieInfo} onSubmit={onSubmit} />
    );

    // Correctly handling numeric inputs to prevent NaN
    fireEvent.change(screen.getByLabelText("Runtime (minutes):"), {
      target: { value: "148" }, // Provide some numeric value
    });

    fireEvent.change(screen.getByLabelText("Rating:"), {
      target: { value: "8.8" }, // Provide some numeric value
    });

    fireEvent.submit(screen.getByRole("button", { name: /submit/i }));

    // Ensure that the expected object aligns with the initial values plus any manual changes done in test
    expect(onSubmit).toHaveBeenCalledWith({
      id: "",
      title: "Inception",
      release_date: "2010-07-16",
      genres: ["Sci-Fi"],
      overview: "A movie about dreams.",
      runtime: 148,
      vote_average: 8.8,
      poster_path: "", // Assuming default or empty values are ""
    });
  });
});
