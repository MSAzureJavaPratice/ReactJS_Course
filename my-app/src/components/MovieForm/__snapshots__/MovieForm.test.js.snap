// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="movie-form"
  >
    <div>
      <label>
        Title:
        <input
          name="title"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Release Date:
        <input
          name="releaseDate"
          type="date"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Genre:
        <select
          multiple=""
          name="genre"
        >
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Crime"
          >
            Crime
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Sci-Fi"
          >
            Sci-Fi
          </option>
          <option
            value="Thriller"
          >
            Thriller
          </option>
          <option
            value="Fantasy"
          >
            Fantasy
          </option>
          <option
            value="Adventure"
          >
            Adventure
          </option>
          <option
            value="Romance"
          >
            Romance
          </option>
        </select>
      </label>
    </div>
    <div>
      <label>
        Runtime (minutes):
        <input
          name="runtime"
          type="number"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Rating:
        <input
          name="rating"
          step="0.1"
          type="number"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Poster URL:
        <input
          name="posterPath"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Overview:
        <textarea
          name="overview"
          rows="5"
        />
      </label>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
